modules/blockbuilder/blockbuilder.go

```diff
@@ -507,33 +510,62 @@ func formatActivePartitions(partitions []int32) string {
 // end record offset. Based on that it sort the partitions by lag
 func (b *BlockBuilder) fetchPartitions(ctx context.Context, partitions []int32) ([]partitionState, error) {
 	var (
-		ps    = make([]partitionState, 0, len(partitions))
-		group = b.cfg.IngestStorageConfig.Kafka.ConsumerGroup
-		topic = b.cfg.IngestStorageConfig.Kafka.Topic
+		ps          = make([]partitionState, 0, len(partitions))
+		commits     kadm.OffsetResponses
+		endsOffsets kadm.ListedOffsets
+		err         error
 	)
 
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Minute,
+		MaxRetries: 5,
+	})
+	for boff.Ongoing() {
+		commits, endsOffsets, err = b.getPartitionOffsets(ctx)
+		if err == nil {
+			break
+		}
+		retryable := ingest.HandleKafkaError(err, b.kafkaClient.ForceMetadataRefresh)
+		if retryable {
+			boff.Wait()
+		}
+	}
```

Comment:
I don't understand this loop and a similar one in `metrics.go`, if the error is non-retryable it retries it immediately anyway while the back-off is still ongoing?

Reply:
right, I have simplified the logic, and it always retries until there is no error or the max number of attempts is reached